Add MicroMeter Netty support

alesj opened the pull request:
Adding initial Netty bytebuf allocator metrics support from Micrometer.

This currently adds all available / used allocators found in Quarkus code, but ...
... `VertxByteBufAllocator.DEFAULT`, which does not implement `ByteBufAllocatorMetricProvider`.

For `QuarkusMultipartFormUpload.ALLOC` reflection was used
(a) for loading the class in `ReactiveNettyMetricsProvider` -- TODO on why is that?
(b) and for accessing `ALLOC` field -- we could add a public static method if/when we resolve (a)

alesj committed:
Add MicroMeter Netty support

brunobat commented on extensions/micrometer/deployment/src/test/java/io/quarkus/micrometer/deployment/binder/NettyMetricsTest.java:
```diff
@@ -0,0 +1,45 @@
+import io.micrometer.core.instrument.binder.netty4.NettyEventExecutorMetrics;
+import io.quarkus.test.QuarkusUnitTest;
+
+public class NettyMetricsTest {
```
We should have tests that cover each one of the classes being instrumented...

alesj replied:
Which classes and instrumented how?

brunobat replied:
How do you know you are getting metrics from the pools and they reflect correct values?
By creating requests on the right places, it should be possible to assert the executor metrics, no?

alesj replied:
> How do you know you are getting metrics from the pools and they reflect correct values?

Atm I trust the Micormeter code do do that.

Let me try and add something ...

brunobat replied:
Thanks @alesj, @franz1981

franz1981 commented on extensions/micrometer/runtime/src/main/java/io/quarkus/micrometer/runtime/export/NettyMetricsProvider.java:
```diff
@@ -0,0 +1,26 @@
+    @Produces
+    @Singleton
+    public MeterBinder unpooledByteBufAllocatorMetrics() {
+        return new NettyAllocatorMetrics(UnpooledByteBufAllocator.DEFAULT);
```
Ok I see they made it right -> https://github.com/micrometer-metrics/micrometer/blob/f6f8eaed9fe7ff1399964bf85b7b850ad0732187/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/netty4/NettyAllocatorMetrics.java#LL52C93-L52C93

ebullient replied:
ok. the binder should not go in runtime/export .. it should go in runtime/binder (just as it does in the micrometer repo)

ebullient commented on extensions/micrometer/runtime/src/main/java/io/quarkus/micrometer/runtime/export/VertxNettyEventExecutorMetricsProvider.java:
move to runtime/binder

alesj commented:
OK, moved as suggested by @ebullient .
Plus added "values" tests as suggested by @brunobat .

franz1981 commented on extensions/micrometer/deployment/src/test/java/io/quarkus/micrometer/deployment/binder/NettyMetricsTest.java:
```diff
@@ -0,0 +1,150 @@
+        Assertions.assertTrue(pbba_found);
+        Assertions.assertTrue(unpbba_found);
+        Assertions.assertTrue(vx_pbba_found);
+        Assertions.assertTrue(vx_unpbba_found);
```
If the client request is the very first one happening from where the server is started you could assert that the used memory values have changed before/after too

alesj committed:
Add memory value check, pending tasks check.
Check if exists / found
Do await in finally
Fill event loop until there are some pending tasks.
Get the value directly from the gauge.
Use more stable test.

alesj commented on extensions/micrometer/deployment/src/test/java/io/quarkus/micrometer/deployment/binder/NettyMetricsTest.java:
```diff
@@ -0,0 +1,150 @@
+                List<Meter> meters = registry.getMeters();
+                for (Meter meter : meters) {
+                    List<Tag> tags = meter.getId().getTags();
+                    if (tags.contains(tag)) {
```
@franz1981 just added this ... but it fails ... 0 pending tasks ... 
I think I got your suggestion right though. Can you check?

franz1981 commented:
@alesj I made an only Netty reproducer

```java
    public static void main(String[] args) throws Exception {
        // by default it would size things with Runtime::getRuntime::availableProcessors
        int threads = Runtime.getRuntime().availableProcessors();;
        EventLoopGroup group = new NioEventLoopGroup(threads);
        int tasks = threads;
        final CyclicBarrier allPendingTasksAreIn = new CyclicBarrier(tasks + 1);
        CountDownLatch waitCollectingMeasures = new CountDownLatch(1);
        List<Future<Future<?>>> pendingTasksCompleted = new ArrayList<>(tasks);
        for (EventExecutor eventLoop : group) {
            pendingTasksCompleted.add(eventLoop.submit(() -> {
                try {
                    Future<?> pendingTask =eventLoop.submit(() -> {});
                    // this executor will have 1 pending task because is still running this current one
                    allPendingTasksAreIn.await();
                    waitCollectingMeasures.await();
                    return pendingTask;
                } catch (Throwable ignore) {
                    return null;
                }

            }));
        }
        allPendingTasksAreIn.await();
        // this would return 0 for everyone
        for (EventExecutor eventLoop : group) {
            int pendingTasks = ((SingleThreadEventLoop) eventLoop).pendingTasks();
            System.out.println("pendingTasks on " + eventLoop + " are: " + pendingTasks);
        }
        waitCollectingMeasures.countDown();
        for (Future<Future<?>> pendingTaskCompleted : pendingTasksCompleted) {
            pendingTaskCompleted.get().get();
        }
        // this would return 0 for everyone
        for (EventExecutor eventLoop : group) {
            int pendingTasks = ((SingleThreadEventLoop) eventLoop).pendingTasks();
            System.out.println("pendingTasks on " + eventLoop + " are: " + pendingTasks);
        }
        group.shutdownGracefully();
    }
```
What it does is to submit to the same event loop while the current executing task is blocking draining more tasks: which allow to have all event loops in a steady state where we can collect the pending tasks metrics.
And await till everything will be drained right after and verify that the number of pending tasks return back to 0.
For Quarkus/Netty we need to make sure no timers (the vertx thread checker @cescoffier  AFAIK) would try to add more tasks. I undertstand that is a very fragile/brittle kind of test, but I cannot think anything better to be sure that suddenly any Netty/Vertx change won't break the observability promises of this feature (which should be tested on Micrometer repo too, TBH).

franz1981 commented on extensions/micrometer/deployment/src/test/java/io/quarkus/micrometer/deployment/binder/NettyMetricsTest.java:
```diff
@@ -0,0 +1,229 @@
+    }
+
+    private void checkMetrics(List<Meter> meters, EventExecutor executor, int expected) {
+        if (executor instanceof SingleThreadEventExecutor) {
```
I would make it fail if it won't match or we risk that a change in the class type would break silently

alesj replied:
Well, metrics are only added for this type ... so I guess we should only check for metrics for this type?

```
    public NettyEventExecutorMetrics(Iterable<EventExecutor> eventExecutors) {
        this.eventExecutors = eventExecutors;
    }

    @Override
    public void bindTo(MeterRegistry registry) {
        this.eventExecutors.forEach(eventExecutor -> {
            if (eventExecutor instanceof SingleThreadEventExecutor) {
```

alesj commented on extensions/micrometer/deployment/src/test/java/io/quarkus/micrometer/deployment/binder/NettyMetricsTest.java:
```diff
@@ -0,0 +1,229 @@
+    }
+
+    @Test
+    public void testEventExecutorMetricsValues() throws Exception {
```
Added ... 1min OK?

alesj commented:
@franz1981 @ebullient ready to merge this?
(@cescoffier)

brunobat commented on extensions/micrometer/runtime/src/main/java/io/quarkus/micrometer/runtime/config/NettyConfigGroup.java:
```diff
@@ -0,0 +1,35 @@
+    public Optional<Boolean> enabled;
+
+    @Override
+    public Optional<Boolean> getEnabled() {
```
I wonder if we should add some sort of documentation to this cool feature.

alesj replied:
Ah, good catch -- yeah, we should definitely add something.
I'll check what we currently have, and add something similar for this / Netty.

franz1981 reviewed (commented):
I didn't checked myself, but the unpooled allocators report any metrics? Afaik nope (for vertx, at least @vietj can confirm)

brunobat commented:
I think this is a release noteworthy feature.

brunobat commented:
We are still missing a docs update

alesj commented:
> We are still missing a docs update

All I could find is this:
* https://quarkus.io/guides/telemetry-micrometer#configuration-reference

And no other specific info/docs on what we support ootb, etc ...

I guess we can merge this one then, and create a new one with ^^ ?

brunobat commented:
> 



> > We are still missing a docs update
> 
> All I could find is this:
> 
> * https://quarkus.io/guides/telemetry-micrometer#configuration-reference
> 
> And no other specific info/docs on what we support ootb, etc ...
> 
> I guess we can merge this one then, and create a new one with ^^ ?

I guess the config you adding will also show up in there.
However I think we should explain somehow what these metrics mean somewhere. If not on the guide maybe a blog post?

brunobat commented:
The metrics are getting out alright.
Just a question. I did a collection of `netty.allocator.memory.used` metrics.
We have 8 of them being generated for multiple combinations of:
* UnpooledByteBufAllocator, direct
* UnpooledByteBufAllocator, heap
* PooledByteBufAllocator, direct
* PooledByteBufAllocator, heap
For different data point Ids

Would it be possible to have something else apart of the number to understand what they mean?
Here is a file with an output example:
[metrics.txt](https://github.com/quarkusio/quarkus/files/12160611/metrics.txt)

alesj commented:
> Would it be possible to have something else apart of the number to understand what they mean? 

This should be raised within Micrometer project -- as that's where the metric collecting code "lives" ...
* https://github.com/micrometer-metrics/micrometer/tree/main/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/netty4

brunobat reviewed (approved):
This looks good!
Let's merge and figure out the documentation later.

brunobat commented:
@alesj can we try to merge this before 3.3?

brunobat commented:
cc @geoand
